### ISSUE TITLE
Workspace assignment fails when translation contains spaces

### ISSUE THREAD

**Issue by jat255** [usnistgov/core_main_app#11]
Hi guys, I discovered the following issue when working on NexusLIMS-CDCS, and since it's an upstream
bug, I wanted to report it. FYI, this bug report was generated using AI, but I've reviewed it and
I believe it's all correct (the underlying issue and root cause is definitely correct).

# Bug Report: Workspace Assignment Fails When Translation Contains Spaces

**Upstream Project:** usnistgov/core_main_app
**Severity:** High
**Component:** Dashboard workspace assignment functionality
**Affected Versions:** 3.18.0 (likely affects all versions)

## Summary

The workspace assignment feature silently fails when the `record_label` Django translation contains spaces (e.g., "experiment record"). This is caused by a design flaw where translated display labels are reused as CSS class names in JavaScript selectors.

## Impact

- Users cannot assign records to workspaces from the dashboard
- No error message is shown to users (silent failure)
- Workaround requires removing spaces from translations, breaking UI display quality
- Affects any deployment using custom translations with multi-word labels

## Steps to Reproduce

1. **Configure a translation with spaces:**
   ```po
   # locale/en/LC_MESSAGES/django.po
   msgid "record_label"
   msgstr "experiment record"
   ```

2. **Navigate to dashboard:** `/dashboard/records`

3. **Attempt to change workspace:**
   - Click the "Actions" dropdown on any record
   - Click "Change workspace"
   - Select a workspace from the modal
   - Click "Change" button

4. **Observe the failure:**
   - Modal closes
   - Page reloads
   - Workspace is NOT changed
   - No error message shown to user

### Problem Location

**File:** `core_main_app/static/core_main_app/user/js/workspaces/list/modals/assign_workspace.js`

**Lines 5-11:**
```javascript
load_form_change_workspace = function() {
    $("#assign-workspace-modal").modal("show");
    var $recordRow = $(this).parent().parent();

    // Get parent if btn in dropdown  (object id undefined)
    if (!$recordRow.attr("objectid")) $recordRow = $recordRow.parent()
    $('.'+functional_object+'-id').val($recordRow.attr("objectid"));
    // ...
}
```

**Line 40:**
```javascript
assign_workspace = function() {
    var workspace_id = $( "#id_workspaces" ).val().trim();
    $.ajax({
        url : assignWorkspaceUrl,
        type : "POST",
        dataType: "json",
        data : {
            workspace_id: workspace_id,
            document_id: getSelectedDocument()  // Returns empty when selector fails
        },
        // ...
    });
};
```

### The Bug

1. **Template uses translation as CSS class:**
   ```html
   <!-- core_main_app/templates/core_main_app/user/workspaces/list/modals/assign_workspace.html -->
   <input type="hidden" class="{{data.document}}-id"/>
   ```
   Where `data.document` comes from the translated `record_label`.

2. **JavaScript builds invalid selector:**
   When `functional_object = "experiment record"`:
   ```javascript
   var selector = '.' + functional_object + '-id';
   // Results in: ".experiment record-id"
   ```

3. **Invalid CSS selector fails:**
   - `.experiment record-id` is interpreted as:
     - Element with class `experiment`
     - Containing a descendant element of type `record-id`
   - The actual element has classes: `class="experiment record-id"` (two classes)
   - Correct selector should be: `.experiment.record-id`

4. **Silent failure cascades:**
   - Hidden input value is never set
   - `getSelectedDocument()` returns `[undefined]`
   - jQuery drops undefined values during serialization
   - POST request sent with only `workspace_id`, no `document_id`
   - Backend receives empty `document_ids` list
   - Backend loop never executes but returns success anyway
   - Frontend reloads page thinking it succeeded

## Current (Bad) Workaround

Change translations to not contain spaces:

```po
msgid "record_label"
msgstr "experiment-record"  # Forced to use hyphen
```

**Problems with this workaround:**
- UI now shows ugly labels: "My Experiment-Records" instead of "My Experiment Records"
- Requires all deployments to avoid spaces in translations
- Defeats the purpose of i18n/l10n (many languages naturally use multi-word terms)
- Hidden constraint not documented anywhere

### Option 1: Use Separate Identifier for CSS Classes (Recommended)

Don't use translated display labels as CSS identifiers. Use a slug or ID field instead.

**In context dict:**
```python
context = {
    'data': {
        'document': _('record_label'),        # For display: "experiment record"
        'document_slug': 'data',              # For CSS/JS: "data"
    }
}
```

**In template:**
```html
<input type="hidden" class="{{data.document_slug}}-id"/>
```

**In JavaScript:**
```javascript
var functional_object_slug = "{{data.document_slug}}";
var selector = '.' + functional_object_slug + '-id';  // ".data-id"
```

### Option 2: Sanitize Translation for CSS Use

Automatically convert spaces to hyphens only when used as CSS class:

**In template filter:**
```python
# Create a template filter
@register.filter
def as_css_class(value):
    """Convert a string to a valid CSS class name."""
    return value.lower().replace(' ', '-').replace('_', '-')
```

**In template:**
```html
<input type="hidden" class="{{data.document|as_css_class}}-id"/>
```

**In JavaScript:**
```javascript
var functional_object = "{{data.document|as_css_class}}";
```

### Backend Validation Missing

**File:** `core_main_app/views/user/ajax.py`, Line 492

```python
document_ids = request.POST.getlist("document_id[]", [])
# ...
for data_id in document_ids:  # Never executes if empty, but returns success!
    # ...
return HttpResponse(json.dumps({}), content_type="application/javascript")
```

The backend should validate that `document_ids` is not empty and return a proper error if no documents are provided.

**Suggested fix:**
```python
document_ids = request.POST.getlist("document_id[]", [])
workspace_id = request.POST.get("workspace_id", None)

# Validate input
if not document_ids:
    logger.error(f"No document_id provided. POST data: {dict(request.POST)}")
    return HttpResponseBadRequest(
        "No documents were selected. Please select a document and try again."
    )
```

## Files Affected

- `core_main_app/static/core_main_app/user/js/workspaces/list/modals/assign_workspace.js`
- `core_dashboard_common_app/static/core_dashboard_common_app/user/js/get_selected_document.js`
- `core_main_app/templates/core_main_app/user/workspaces/list/modals/assign_workspace.html`
- `core_main_app/views/user/ajax.py` (validation issue)

### Test Case 1: Translation with spaces
```po
msgid "record_label"
msgstr "experiment record"
```
**Expected:** Workspace assignment should work
**Actual:** Silent failure

### Test Case 2: Translation with hyphen
```po
msgid "record_label"
msgstr "experiment-record"
```
**Expected:** Workspace assignment works, but UI labels look bad
**Actual:** Works (current workaround)

### Test Case 3: After fix
**Expected:** Both should work - translations can have spaces, and workspace assignment succeeds

## Environment

- MDCS Version: 3.18.0
- Browser: Firefox 146.0, Chrome (reproduced in both)
- Database: PostgreSQL 17
- Python: 3.11

## Related Code References

1. **Modal template:** `core_main_app/templates/core_main_app/user/workspaces/list/modals/assign_workspace.html`
2. **JavaScript handlers:** `core_main_app/static/core_main_app/user/js/workspaces/list/modals/assign_workspace.js`
3. **Document selector:** `core_dashboard_common_app/static/core_dashboard_common_app/user/js/get_selected_document.js`
4. **Backend view:** `core_main_app/views/user/ajax.py` (AssignView class)
5. **Table template:** `core_dashboard_common_app/templates/core_dashboard_common_app/list/my_dashboard_my_records_table.html`

## Recommendation

**Option 1 (separate slug field)** is the cleanest long-term solution as it:
- Properly separates concerns (display vs. identifiers)
- Fully supports i18n/l10n
- Prevents similar issues in the future
- Follows web development best practices (semantic HTML/CSS)

However, it requires changes across templates, views, and JavaScript.

---

**Additional Notes:**

This issue demonstrates a broader anti-pattern in the codebase: using user-facing translated strings as technical identifiers (CSS classes, HTML IDs, etc.). A comprehensive audit should be performed to find other instances of this pattern and refactor to use separate identifier fields.

**Comment by pdessauw** [comment]
Hi @jat255 

The fix has been pushed and published with version [3.20](https://github.com/usnistgov/MDCS/releases/tag/3.20.0) of the CDCS. The changes have been implemented in `core_dashboard_common_app` and propagated to `core_main_app`, `core_dashboard_app` and `core_dashboard_regitry_app`.

Your comment was incredibly valuable to help debug this issue with the root cause being a [`FUNCTIONAL_OBJECT_ENUM`](https://github.com/usnistgov/core_dashboard_common_app/blob/develop/core_dashboard_common_app/constants.py#L95) using translation both for display and triggering actions. 

We left the original data structure intact and instead used the name (keys) of the enum for triggering actions. The values of this enum are now used for display only.

Here is the list of all commits related to the issue:
* https://github.com/usnistgov/core_dashboard_common_app/commit/7f35a103d590fc098e8e83b70270119d5164df9d
* [usnistgov/core_main_app@`c3f7050`](https://github.com/usnistgov/core_main_app/commit/c3f7050e6dd6891ea377af9cf132b70339fd3546)
* https://github.com/usnistgov/core_dashboard_app/commit/e7699621f908f0db124f4348861df7bf54b22784
* https://github.com/usnistgov/core_dashboard_registry_app/commit/7d22dd1830465933d965e5cf4cbd43695b1fdd7f

Please let us know if you encounter any other issues 😄
